Augment Intro to Tests topic (#151)
Ada-Developers-Academy/core-unit-1: intro-to-tests/automated-tests.md

### Example: Astrology Software

### !callout-info

## Astrological Birth Chart
An Astrological Birth Chart is a visual map of what the sky looked like when you were born. It is determined using your exact birthdate and birthplace. It includes three signs, your **sun sign** (typically refered to as your Zodiac sign), your **moon sign**, and your **rising sign**. 

> [source](https://www.shape.com/lifestyle/mind-and-body/sun-moon-rising-sign-meaning)

Reviewer: Birth charts are pretty complicated and lots of people (like me) don't know anything about them.

Author: I agree, me too. That's why I added this callout to provide some context. Do you think more context is needed to ensure this example is understandable?

### !end-callout
<!-- available callout types: info, success, warning, danger, secondary  -->

![Part of an astrology birth chart](../assets/intro-to-tests_automated-tests_astrology.jpg)  
[(source)](https://pixabay.com/photos/astrology-divination-chart-993127/)

Addison is writing software about astrology. They are working on a feature that, when a user provides a date, time, and location of their birth, will compute their astrological birth chart. They need to know if they've implemented the logic correctly; does the feature accurately compute the astrological birth chart? They can test their feature by finding the following information:

1. Get a sample of at least one correct example: an input and and output

>    - This means,∂ finding and using a real date, time, and location, and its real astrological birth chart

Reviewer: Extra character?

```suggestion
   - This means, finding and using a real date, time, and location, and its real astrological birth chart
```

### Test Cases and Scenarios

> Testing a single function will generally have many unit tests to test it. Why? Based on the context, a single function could potentially have different arguments that give back different return values. Different scenarios in development can imply different arguments, which would produce different logic. All scenarios and possible outcomes of a function should be tested, as much as reasonably possible.

Reviewer: ```suggestion
Testing a single function will generally have many unit tests to ensure the code works correctly. Why? Based on the context, a single function could potentially have different arguments that give back different return values. Different scenarios in development can imply different arguments, which would produce different logic. All scenarios and possible outcomes of a function should be tested, as thoroughly as possible.
```

Addison could come up with the following scenarios:

| Question to Ask                                                             | Specific Arguments                          | Expected Return Value       |
| --------------------------------------------------------------------------- | ------------------------------------------- | --------------------------- |
| 1. Do we get an accurate birth chart if we pass in a date, location, and time? | `"January 1, 1980"`, `"Seattle"`, `"12:00"` | `{"Sun": "Capricorn", ...}` |
| 2. What happens if the function receives `None` for time?                      | `"January 1, 1980"`, `"Seattle"`, `None`    | `None`                      |
| 3. What happens if the `date` argument comes as a different format?            | `"1980-01-01"`, `"Seattle"`, `"12:00"`      | `{"Sun": "Capricorn", ...}` |
| 4. What happens if the function receives an invalid place for `location`?       | `"January 1, 1980"`, `"xyz"`, `"12:00"` | should raise an exception   |

To ensure that our test cases cover all scenarios, it can be helpful to categorize tests using the following categories:
1. Positive Nominal
1. Negative Nominal
1. Positive Edge

> 1. Negative Edge

Reviewer 2: Up to your judgement, but I think we don't need to emphasize the positive and negative here; separation into nominal and edge cases feels sufficient.

#### Nominal Test Cases

> A nominal test case is a type of test case that describes a piece of core functionality needed for the success of this method. This is the test case that verifies that the method does its primary responsibility. Nominal test cases are the obvious test cases.

Reviewer 2: Consider adding that nominal test cases are often referred to as "happy path" test cases.

In the case of Addison's astrological software, the first test case listed above is a nominal **positive nominal** test case. It checks that given the correct, expected input (the `date`, `location`, and `time`), the functions returns the correct output (a dictionary with the sun, rising, and moon signs). 

A **negative nominal** test case is a type of test case that describes the most obvious, most typical way that a function handles unexpected input. Again, consider our test cases outlined above, the second test case falls under this category. If we pass in `None` for time, the function

> does not have enough information to determine the astrological birth chart, and as such, it returns `None`.

Reviewer 2: I've never heard of a negative nominal case, and I'm not seeing it coming up in Google searches either. To me, `None` feels like an edge case, and we can remove discussion of negative nominal entirely.

#### Edge Test Cases

An edge test case is a type of test case that verifies that the function can work successfully, even given non-obvious context. This is the test case that verifies that the function works, even with very unexpected context.

A **positive edge** test case is a test case that describes a set of inputs and expectations that are on the limits of the method's most obvious, most typical way of working successfully. In the test cases listed above, the third test case is a positive edge case. The function returns the correct birth chart even when the input is in an unexpected format.

> A **negative edge** test case is a test case that tests that this function can handle the most non-obvious, most atypical unexpected input gracefully. The fourth example above is a negative edge case.The function receives an invalid place for `location` and raises an exception. We will learn more about raising, handling, and testing exceptions in [Exception Handling](../exception-handling/raising-and-handling-exceptions.md)

Reviewer 2: ```suggestion
A **negative edge** test case is a test case that tests that this function can handle the most non-obvious, most atypical unexpected input gracefully. The fourth example above is a negative edge case. The function receives an invalid place for `location` and raises an exception. We will learn more about raising, handling, and testing exceptions in [Exception Handling](../exception-handling/raising-and-handling-exceptions.md)
```

##### !end-answer

##### !explanation 

> Tests and these benefits are reasons to follow a practice called test-driven development.~

Reviewer 2: ```suggestion
Tests and these benefits are reasons to follow a practice called test-driven development.
```
